lib/core/src/irods_network_manager.cpp

```diff
@@ -1,118 +1,196 @@
         // call generic plugin loader
         network* net = 0;
-        error ret = load_plugin< network >(
+        error ret = load_plugin<network>(
                         net,
                         _plugin_name,
                         KW_CFG_PLUGIN_TYPE_NETWORK,
                         _inst_name,
-                        _context );
-        if ( ret.ok() && net ) {
-            _plugin.reset( net );
+                        _context);
+        if (ret.ok() && (net != nullptr)) {
+            _plugin.reset(net);
             return SUCCESS();
 
         }
         else {
-            return PASS( ret );
-
+            return PASS(ret);
         }
 
     } // load_network_plugin
 
-// =-=-=-=-=-=-=-
-// public - given a type, load up a network plugin
+    // =-=-=-=-=-=-=-
+    // public - return a network plugin, loading if necessary
+    error network_manager::get_plugin(
+        const int&         _proc_type,
+        const std::string& _type,
+        const std::string& _inst,
+        const std::string& _ctx,
+        network_ptr&       _net)
+    {
+        plugin_map_key key(_proc_type, _type, _inst, _ctx);
+
+        // look for existing plugin
+        {
+            // we're going to read the plugins table
+            // acquire shared lock so writers block
+            std::shared_lock lock(table_lock_);
+
+            auto plugin_itr_ = plugins_.find(key);
+            if (plugin_itr_ != plugins_.end()) {
+                _net = plugin_itr_->second;
+                return SUCCESS();
+            }
+        }
+
+        // plugin not loaded yet
+        {
+            // we're (probably) going to write to the plugins table
+            // acquire unique lock so other accessors block
+            std::unique_lock lock(table_lock_);
+
+            // check table again in case the plugin was added while we were waiting on the lock
+            auto plugin_itr_ = plugins_.find(key);
+            if (plugin_itr_ != plugins_.end()) {
+                _net = plugin_itr_->second;
+                return SUCCESS();
+            }
+
+            std::string type = _type;
+            type += (CLIENT_PT == _proc_type) ? "_client" : "_server";
+
+            // load plugin
+            network_ptr ptr;
+            error ret = load_network_plugin(ptr, type, _inst, _ctx);
+            if (!ret.ok()) {
+                return PASSMSG("Failed to load network plugin", ret);
+            }
+
+            // insert plugin into table
+            bool was_inserted = false;
+            std::tie(plugin_itr_, was_inserted) = plugins_.try_emplace(key, ptr);
+            if (!was_inserted) {
+                return ERROR(SYS_INTERNAL_ERR, "Plugin already in table despite lock");
+            }
+
+            _net = plugin_itr_->second;
+
+            // insert into legacy plugins table (for deprecated functions)
+            ret = plugins_old_.set(_inst, ptr);
+            if (!ret.ok()) {
+                log(LOG_WARNING, ret.result());
```

Comment:
Is this error case a serious enough problem to return something other than `SUCCESS()`?

Reply:
Also... I know we are doing this in other places in the client libraries/plugins, but is it a problem to try to write to the log here? I seem to recall @korydraughn cautioning against this, but I don't remember why, or whether I'm thinking of something else.

Reply:
Use of the logger must be handled with care here due to the network plugins being used by the clients and server.

Reply:
> Is this error case a serious enough problem to return something other than `SUCCESS()`?

I don't think so. As long as the deprecated functions aren't being used, this kind of failure would never cause a problem.

I can remove the logging here if needed.

Reply:
Opened #7779 to cover the TODO.